### PR TITLE
maint: merge dependency specifications into setup.cfg

### PR DESCRIPTION
**Pull request by neutrinoceros**
## PR Summary

This is a second step forward in trying to solve #3074, following #3078.
Goal: remove as much content as possible from
- `tests/test_minimal_requirements.txt`
- `tests/test_requirements.txt`
- and`tests/test_prerequirements.txt`

and migrate it to setup.cfg
The migration process is error prone and I expect a few iterations will be required.

TODO
- [x] define a `test` target
- [x] cleanup duplicated content
- [x] define a `minimal` target
- [x] cleanup duplicated content
- [x] check that the overlap between minimal req and requirements leads to a correct env (e.g if we mix `matplotlib>=2.0.2` with `matplotlib==2.0.2`, do we actually satisfy that last requirement ?)
- [x] define a `full` target (corresponding to `test_requirements.txt`
- [x] cleanup duplicated content
- [x] check `pyyaml` VS `pyaml` (both are currently "required" from test_requirements.txt but I bet that's a useless redundancy)
- [x] cleanup (self review)
- [x] reactivate the pre-commit hook (can be done at any point but I had problems with it in my first attempt)
- [x] reviews
- [ ] deal with the Jenkins side of things

This last step can be performed with a conditional similar to
```
if [ -f test_requirements.txt ] ; then
    # old strategy
else
   # new strategy (replicate what I'm doing in ci_install.sh
fi
```

### Cleanup yaml deps.
Currently minimal build requires `pyyaml`, but we use `pyaml` (which itself _depends_ on `pyyaml`) in the other test envs. Note that both import as `yaml`... Furtermore, `yaml` is a soft dependency: part of `on_demand_imports`, tests use `@requires_module("yaml")`), _but_ there's actually one place in the code base that we import it without going through on_demand_imports.py : https://github.com/yt-project/yt/blob/16cf684ca50a3ad034e8721b184a7609dd192772/yt/utilities/mesh_code_generation.py#L1
so I'd be inclined to think that it's de-facto already a hard dependency ?

### Cleanup pre-existing env targets ?
see https://github.com/yt-project/yt/blob/16cf684ca50a3ad034e8721b184a7609dd192772/setup.cfg#L61
apparently the `hub` target only adds one requirement (`girder_hub`) and is only used in parts of the command line tool. I'm not sure atm wether this is still used or tested for.

Same goes for the `bottle` target.

update : according to Kacper the first one is still used.

### Remove `tests/windows_conda_requirements.txt`
I introduced this file in the present PR as an intermediate step, though I didn't manage to suppress it completely yet.

### Remove `tests/cartopy_requirements.txt`
This is a convenience to mitigate problems with cartopy+shapely distribution. I hope it can be removed in the future, but that will probably require upstream changes regarding packaging.

### CODE REVIEW

**Comment by neutrinoceros** [comment]
Since this isn't going to get merged for at least a couple of days, I'll take the opportunity to experiment a little bit further than I originally intended

**Review by matthewturk — APPROVED** [review]
This looks like an improvement to me.  Anyone know if this will impact things like using `repo2docker` etc?

File: tests/ci_install.sh [hunk 1]
```diff
@@ -30,43 +30,34 @@ cp tests/matplotlibrc .
...
-    H5PY=$(grep h5py tests/test_requirements.txt)
-    MATPLOTLIB=$(grep matplotlib tests/test_requirements.txt)
-    SCIPY=$(grep scipy tests/test_requirements.txt)
-    conda config --set always_yes yes
-    conda info -a
-    conda install --quiet --yes -c conda-forge \
-      $CYTHON $NUMPY $CARTOPY $H5PY $MATPLOTLIB $SCIPY
+    # windows_conda_requirements.txt is a survivance of test_requirements.txt
+    # keep in sync: setup.cfg
+    while read requirement; do conda install --yes $requirement; done < tests/windows_conda_requirements.txt
```
Comment (cphyc) [inline tests/ci_install.sh:35]:
```suggestion
    conda install --yes --file tests/windows_conda_requirements.txt
```

File: tests/ci_install.sh [hunk 2]
```diff
@@ -30,43 +30,34 @@ cp tests/matplotlibrc .
...
-# Step 2: install required packages (depending on whether the build is minimal)
+# Step 2: install deps and yt
 if [[ ${dependencies} == "minimal" ]]; then
-    # Ensure numpy and cython are installed so dependencies that need to be built
-    # don't error out
-    # The first numpy to support py3.6 is 1.12, but numpy 1.13 matches
-    # unyt so we'll match it here.
-    python -m pip install numpy==1.13.3 cython==0.26.1
-    python -m pip install -r tests/test_minimal_requirements.txt
+    python -m pip install -e .[test,minimal]
```
Comment (cphyc) [inline tests/ci_install.sh:44]: So much better!!

**Review by cphyc — APPROVED** [review]
I'm very happy about this! You have my green light once the tests pass

File: tests/ci_install.sh (hunk 1 above)
Comment (neutrinoceros) [reply tests/ci_install.sh:35]: I don't think it works unfortunately. To the best of my knowledge, Conda can only install an named environment from a yaml file, and I didn't want to invest any more effort in a temporary band aid

File: tests/ci_install.sh [hunk 3]
```diff
@@ -30,43 +30,34 @@ cp tests/matplotlibrc .
...
-    # These versions are pinned, so we will need to update/remove them when
-    # the hack is no longer necessary.
-    python -m pip install -r tests/test_prerequirements.txt
-    CFLAGS="$CFLAGS -DACCEPT_USE_OF_DEPRECATED_PROJ_API_H" python -m pip install -r tests/test_requirements.txt
-fi
+    # Cython and numpy as build-time requirements to the following optional deps in yt
+    # - cartopy
+    # - netcdf4
+    # - pyqt5
+    # The build system are however not specified properly in these projects at the moment
```
Comment (neutrinoceros) [inline tests/ci_install.sh:50]:
```suggestion
    # The build system is however not specified properly in these projects at the moment
```

File: tests/ci_install.sh [hunk 4]
```diff
@@ -30,43 +30,34 @@ cp tests/matplotlibrc .
...
 else
-    # Getting cartopy installed requires getting cython and numpy installed
-    # first; this is potentially going to be fixed with the inclusion of
-    # pyproject.toml in cartopy.
-    # These versions are pinned, so we will need to update/remove them when
-    # the hack is no longer necessary.
-    python -m pip install -r tests/test_prerequirements.txt
-    CFLAGS="$CFLAGS -DACCEPT_USE_OF_DEPRECATED_PROJ_API_H" python -m pip install -r tests/test_requirements.txt
-fi
+    # Cython and numpy as build-time requirements to the following optional deps in yt
```
Comment (neutrinoceros) [inline tests/ci_install.sh:46]:
```suggestion
    # Cython and numpy are build-time requirements to the following optional deps in yt
```

**Review by neutrinoceros — COMMENTED** [review]
self review to correct typos

**Comment by Xarthisius** [comment]
@yt-fido test this please

**Comment by neutrinoceros** [comment]
Thanks @Xarthisius, looks like it worked for the answer test job, and likely for the py38 job as well. The docs and py38_unitests failed however. If there's anything I can do on this size to fix it let me know !

**Comment by Xarthisius** [comment]
> Thanks @Xarthisius, looks like it worked for the answer test job, and likely for the py38 job as well. The docs and py38_unitests failed however. If there's anything I can do on this size to fix it let me know !

It's just me being clumsy. It will probably take a couple of iterations until I get it right...

**Comment by Xarthisius** [comment]
BTW there are extra dependencies for docs that could be added to the setup:

```
pip install pyregion Sphinx==3.1.2 alabaster sphinx-bootstrap-theme sphinx-rtd-theme RunNotebook
pip install -U nbconvert==5.6.1
```

I don't remember specific bugs that resulted in pinning Sphinx and nbconvert, but they're there for a reason I'm sure.

**Comment by neutrinoceros** [comment]
> It's just me being clumsy. It will probably take a couple of iterations until I get it right...

It took a lot more iterations to me, I get it :)

> BTW there are extra dependencies for docs that could be added to the setup

Good to know, I'll make sure to include this (back in a moment).

**Comment by neutrinoceros** [comment]
@Xarthisius does this work for you ? 
https://github.com/yt-project/yt/pull/3236/commits/69a262bc479c03d9c90294cd6222126d8238c474

**Comment by Xarthisius** [comment]
> @Xarthisius does this work for you ?
> [69a262b](https://github.com/yt-project/yt/commit/69a262bc479c03d9c90294cd6222126d8238c474)

LGTM. Thanks!

**Comment by Xarthisius** [comment]
@yt-fido test this please

**Comment by Xarthisius** [comment]
@yt-fido test this please

**Comment by Xarthisius** [comment]
@neutrinoceros I'm doing `python -m pip install -e '.[test,full,docs]'` and it doesn't install sphinx. Any hints?

**Comment by Xarthisius** [comment]
@yt-fido test this please

**Comment by neutrinoceros** [comment]
> @neutrinoceros I'm doing python -m pip install -e '.[test,full,docs]' and it doesn't install sphinx. Any hints?
Maybe, because it's doc not docs...

ah well, that's a good reason :)
I'm surprised that it doesn't give better hints on this kind of error though.

**Comment by neutrinoceros** [comment]
@Xarthisius Am I crazy or is the  failure on the py38 job new ?

**Comment by Xarthisius** [comment]
@yt-fido test this please

**Comment by neutrinoceros** [comment]
I really don't get the what's happening in the one test that still breaks on Jenkins.
@Xarthisius can you show me the scripts you're using there to install the deps, both old and new ?

Also I think your setup for the docs job is likely ok now, but we ran into a timeout two times in a row... (third time's the charm ?)

@yt-fido test this please

**Comment by neutrinoceros** [comment]
Quick note that although the remaining failure happens consistently, the quantitative error that's produced fluctuates.

**Comment by neutrinoceros** [comment]
@yt-fido test this please

**Comment by neutrinoceros** [comment]
@Xarthisius I just performed some diffs between a reference run and the latest one here for the failing docs job on Jenkins and an important difference is that on this branch, we apparently install yt twice, as the log contains
```bash
python -m pip install '.[test,full,doc]'
...
python -m pip install -e '.[test,full,doc]'
```
while in the ref run, it looks to me that yt itself is never *installed* but is only importable because we run
```bash
python setup.py build_ext -i -f
```
and every test is run from the top level of the repo.
I don't know if this sole difference can account for more than an hour of additional overhead though, but there seems to be room for optimisation there (install just once, and probably not in editable mode).
Any chance you already found something more interesting ? :)

**Comment by Xarthisius** [comment]
Doc build failed due to timeout. It happens from time to time when sphinx subthread dies for some reason. The build is kind brittle this way. I'd give another chance (which I did a while ago).

**Comment by neutrinoceros** [comment]
oh I see now that it's hanging again against the same volume rendering recipe that it did earlier today. The timeout already seemed to happen consistently on this branch.

**Comment by Xarthisius** [comment]
The errors I'm seeing in the log (that are causing the timeout):

```
Traceback (most recent call last):
  File "temp.py", line 3, in <module>
    ds = yt.load("gadget_cosmology_plus/snap_N128L16_132.hdf5")
  File "/var/jenkins_home/workspace/yt_docs_new/yt/loaders.py", line 81, in load
    fn = str(lookup_on_disk_data(fn))
  File "/var/jenkins_home/workspace/yt_docs_new/yt/sample_data/api.py", line 168, in lookup_on_disk_data
    raise FileNotFoundError(err_msg)
FileNotFoundError: No such file or directory: 'gadget_cosmology_plus/snap_N128L16_132.hdf5'.
```

```
Traceback (most recent call last):
  File "temp.py", line 10, in <module>
    profile = yt.create_profile(
  File "/var/jenkins_home/workspace/yt_docs_new/yt/data_objects/profiles.py", line 1494, in create_profile
    obj.set_field_unit(field, unit)
  File "/var/jenkins_home/workspace/yt_docs_new/yt/data_objects/profiles.py", line 121, in set_field_unit
    fd = self.field_map[field]
KeyError: ('gas', 'cell_mass')
```

**Comment by Xarthisius** [comment]
How about rebasing onto main if you haven't already?

**Comment by neutrinoceros** [comment]
> How about rebasing onto main if you haven't already?

that what I did a few hours ago

**Comment by neutrinoceros** [comment]
This errors you mentioned also happen on a successful run. See for instance: https://tests.yt-project.org/job/yt_docs_new/1508/consoleFull

**Comment by Xarthisius** [comment]
The reason it fails is this:

```
python: LineString.cpp:121: const geos::geom::CoordinateSequence* geos::geom::LineString::getCoordinatesRO() const: Assertion `0 != points.get()' failed.                                      

Exception occurred:
  File "/var/jenkins_home/shiningpanda/jobs/c9e5cb26/virtualenvs/d41d8cd9/lib/python3.8/site-packages/nbconvert/preprocessors/execute.py", line 510, in _check_alive
    raise DeadKernelError("Kernel died")
nbconvert.preprocessors.execute.DeadKernelError: Kernel died
```

we get that error if `shapely` is not compiled with the version of GEOS library present on the testing box. In `requirements.txt` it's taken care of by `--no-binary` flag. We need an equivalent for setup.py provided it exists...

**Comment by neutrinoceros** [comment]
@yt-fido test this please

**Review by Xarthisius — APPROVED** [review]

**Comment by neutrinoceros** [comment]
Hurray ! Many thanks to @Xarthisius for your help with this one !

**Pull request merged**